## Importing the Libraries 

In [34]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split


np.random.seed(1)

## Importing the LAUS dataset

In [35]:
df = pd.read_csv('C:/Users/Srinidhi/Documents/USF/Data_Science_Programming/week7_assignment/Local_Area_Unemployment_Statistics__LAUS_.csv')
df.head(5)

Area Type                           Area Name         Date  Year   \
0              State                          California  01-01-1976   1976   
1              State                          California  01-01-1976   1976   
2             County                  Los Angeles County  01-01-1976   1976   
3             County                  Los Angeles County  01-01-1976   1976   
4  Metropolitan Area  Los Angeles-Long Beach-Glendale MD  01-01-1976   1976   

     Month Seasonally Adjusted (Y/N)  Status (Preliminary / Final)   \
0  January                          N                         Final   
1  January                          Y                         Final   
2  January                          N                         Final   
3  January                          Y                         Final   
4  January                          N                         Final   

   Labor Force  Employment   Unemployment   Unemployment Rate  
0      9672362      8668016        1004346                  0  
1      9774280      8875685         898595                  0  
2      3364151      3040058         324093                  0  
3      3381856      3081806         300050                  0  
4      3364151      3040058         324093                  0

## Goal and the problem Statement

Here based on the dataset alone, I will be considering the Unemployment Rate to be my target variable. My goal is to determine the unemployment rate based on year and the area where it exist. But the dependent variable is continuous in nature. Since we are not working with the RMSE and regression I have chosen to consider the values which lies between 0 and 0.5 as one class and the values between 0.5 to 1 as the other class.

## Finding out the Data type for each column

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194346 entries, 0 to 194345
Data columns (total 11 columns):
 #   Column                         Non-Null Count   Dtype 
---  ------                         --------------   ----- 
 0   Area Type                      194346 non-null  object
 1   Area Name                      194346 non-null  object
 2   Date                           194346 non-null  object
 3   Year                           194346 non-null  int64 
 4   Month                          194346 non-null  object
 5   Seasonally Adjusted (Y/N)      194346 non-null  object
 6   Status (Preliminary / Final)   194346 non-null  object
 7   Labor Force                    194346 non-null  int64 
 8   Employment                     194346 non-null  int64 
 9   Unemployment                   194346 non-null  int64 
 10  Unemployment Rate              194346 non-null  int64 
dtypes: int64(5), object(6)
memory usage: 16.3+ MB


As per the above output, it can be seen that there are different dtypes in the dataset. 

## Finding the nulls

In [37]:
df.isnull().sum()

Area Type                        0
Area Name                        0
Date                             0
Year                             0
Month                            0
Seasonally Adjusted (Y/N)        0
Status (Preliminary / Final)     0
Labor Force                      0
Employment                       0
Unemployment                     0
Unemployment Rate                0
dtype: int64

There are no null values in this dataset.

## Listing out the column names to make sure the right format is utilized later in the cleaning process.

In [38]:
print(df.columns.tolist())

['Area Type ', 'Area Name ', 'Date', 'Year ', 'Month', 'Seasonally Adjusted (Y/N) ', 'Status (Preliminary / Final) ', 'Labor Force', 'Employment ', 'Unemployment ', 'Unemployment Rate']


## Finding if there are any duplicate values.

In [39]:
df.duplicated().sum()

0

In [40]:
df.describe().round(2)

Year   Labor Force  Employment   Unemployment   Unemployment Rate
count  194346.00    194346.00    194346.00      194346.00          194346.00
mean     2013.36    252124.52    234108.95       18015.30               0.00
std         7.51   1361094.53   1264398.57      102912.19               0.05
min      1976.00         0.00         0.00           0.00               0.00
25%      2011.00      1700.00      1572.50         100.00               0.00
50%      2015.00      9600.00      8800.00         700.00               0.00
75%      2019.00     52800.00     47900.00        3979.00               0.00
max      2023.00  19702600.00  18872000.00     3002100.00               1.00

## Finding the uniqueness in the values for different variables.

In [41]:
len(df['Area Name '].unique())

1050

In [42]:
df.sample(5)

Area Type               Area Name         Date  Year     Month  \
143095  Sub-County Place           Corcoran city  10-01-2014   2014  October   
116521            County        Calaveras County  08-01-2018   2018   August   
146137  Sub-County Place              Chino city  10-01-2017   2017  October   
113933  Sub-County Place              Keeler CDP  08-01-2015   2015   August   
183948    Metro Division  San Rafael, CA Met Div  04-01-2016   2016    April   

       Seasonally Adjusted (Y/N)  Status (Preliminary / Final)   Labor Force  \
143095                          N                         Final         4500   
116521                          N                         Final        21490   
146137                          N                         Final        42000   
113933                          N                         Final           40   
183948                          Y                    04-01-2016       139578   

        Employment   Unemployment   Unemployment Rate  
143095         3900            500                  0  
116521        20630            850                  0  
146137        40400           1600                  0  
113933           40              0                  0  
183948       134974           4604                  0

In [43]:
len(df['Area Type '].unique())

6

In [44]:
len(df['Status (Preliminary / Final) '].unique())

399

In [45]:
len(df['Unemployment Rate'].unique())

2

## Dropping the values

Since I am working with the year alone and the month, I will be dropping the date column from the dataset.

In [46]:
df_new= df.drop(['Date'], axis=1)

In [47]:
df_new.head()

Area Type                           Area Name   Year     Month  \
0              State                          California   1976  January   
1              State                          California   1976  January   
2             County                  Los Angeles County   1976  January   
3             County                  Los Angeles County   1976  January   
4  Metropolitan Area  Los Angeles-Long Beach-Glendale MD   1976  January   

  Seasonally Adjusted (Y/N)  Status (Preliminary / Final)   Labor Force  \
0                          N                         Final      9672362   
1                          Y                         Final      9774280   
2                          N                         Final      3364151   
3                          Y                         Final      3381856   
4                          N                         Final      3364151   

   Employment   Unemployment   Unemployment Rate  
0      8668016        1004346                  0  
1      8875685         898595                  0  
2      3040058         324093                  0  
3      3081806         300050                  0  
4      3040058         324093                  0

Further processing the data

In [48]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194346 entries, 0 to 194345
Data columns (total 10 columns):
 #   Column                         Non-Null Count   Dtype 
---  ------                         --------------   ----- 
 0   Area Type                      194346 non-null  object
 1   Area Name                      194346 non-null  object
 2   Year                           194346 non-null  int64 
 3   Month                          194346 non-null  object
 4   Seasonally Adjusted (Y/N)      194346 non-null  object
 5   Status (Preliminary / Final)   194346 non-null  object
 6   Labor Force                    194346 non-null  int64 
 7   Employment                     194346 non-null  int64 
 8   Unemployment                   194346 non-null  int64 
 9   Unemployment Rate              194346 non-null  int64 
dtypes: int64(5), object(5)
memory usage: 14.8+ MB


In [49]:
df_new['Seasonally Adjusted (Y/N)_new '] = df_new['Seasonally Adjusted (Y/N) '].map({'N': 0, 'Y': 1})


In [50]:
df_new

Area Type                           Area Name   Year   \
0                   State                          California   1976   
1                   State                          California   1976   
2                  County                  Los Angeles County   1976   
3                  County                  Los Angeles County   1976   
4       Metropolitan Area  Los Angeles-Long Beach-Glendale MD   1976   
...                   ...                                 ...    ...   
194341                MSA                   Yuba City, CA MSA   2022   
194342                MSA                   Yuba City, CA MSA   2022   
194343                MSA                   Yuba City, CA MSA   2022   
194344                MSA                   Yuba City, CA MSA   2022   
194345                MSA                   Yuba City, CA MSA   2023   

            Month Seasonally Adjusted (Y/N)  Status (Preliminary / Final)   \
0         January                          N                         Final   
1         January                          Y                         Final   
2         January                          N                         Final   
3         January                          Y                         Final   
4         January                          N                         Final   
...           ...                        ...                           ...   
194341  September                          Y                    09-01-2022   
194342    October                          Y                    10-01-2022   
194343   November                          Y                    11-01-2022   
194344   December                          Y                    12-01-2022   
194345    January                          Y                    01-01-2023   

        Labor Force  Employment   Unemployment   Unemployment Rate  \
0           9672362      8668016        1004346                  0   
1           9774280      8875685         898595                  0   
2           3364151      3040058         324093                  0   
3           3381856      3081806         300050                  0   
4           3364151      3040058         324093                  0   
...             ...          ...            ...                ...   
194341        77047        72379           4668                  0   
194342        77065        72185           4880                  0   
194343        77023        71990           5033                  0   
194344        76942        71942           5000                  0   
194345        77249        72170           5079                  0   

        Seasonally Adjusted (Y/N)_new   
0                                    0  
1                                    1  
2                                    0  
3                                    1  
4                                    0  
...                                ...  
194341                               1  
194342                               1  
194343                               1  
194344                               1  
194345                               1  

[194346 rows x 11 columns]

In [51]:
df_new1= df_new.drop(['Seasonally Adjusted (Y/N) '], axis=1)

In [52]:
df_new1

Area Type                           Area Name   Year   \
0                   State                          California   1976   
1                   State                          California   1976   
2                  County                  Los Angeles County   1976   
3                  County                  Los Angeles County   1976   
4       Metropolitan Area  Los Angeles-Long Beach-Glendale MD   1976   
...                   ...                                 ...    ...   
194341                MSA                   Yuba City, CA MSA   2022   
194342                MSA                   Yuba City, CA MSA   2022   
194343                MSA                   Yuba City, CA MSA   2022   
194344                MSA                   Yuba City, CA MSA   2022   
194345                MSA                   Yuba City, CA MSA   2023   

            Month Status (Preliminary / Final)   Labor Force  Employment   \
0         January                         Final      9672362      8668016   
1         January                         Final      9774280      8875685   
2         January                         Final      3364151      3040058   
3         January                         Final      3381856      3081806   
4         January                         Final      3364151      3040058   
...           ...                           ...          ...          ...   
194341  September                    09-01-2022        77047        72379   
194342    October                    10-01-2022        77065        72185   
194343   November                    11-01-2022        77023        71990   
194344   December                    12-01-2022        76942        71942   
194345    January                    01-01-2023        77249        72170   

        Unemployment   Unemployment Rate  Seasonally Adjusted (Y/N)_new   
0             1004346                  0                               0  
1              898595                  0                               1  
2              324093                  0                               0  
3              300050                  0                               1  
4              324093                  0                               0  
...               ...                ...                             ...  
194341           4668                  0                               1  
194342           4880                  0                               1  
194343           5033                  0                               1  
194344           5000                  0                               1  
194345           5079                  0                               1  

[194346 rows x 10 columns]

The above is the data after further processing. Next we shall be encoding the required information using one hot encoding.

## One hot encoding the variables

In [53]:
EncodedData = pd.get_dummies(df_new1, columns = ['Area Type ', 'Area Name ', 'Status (Preliminary / Final) ', 'Month'])
EncodedData

Year   Labor Force  Employment   Unemployment   Unemployment Rate  \
0        1976      9672362      8668016        1004346                  0   
1        1976      9774280      8875685         898595                  0   
2        1976      3364151      3040058         324093                  0   
3        1976      3381856      3081806         300050                  0   
4        1976      3364151      3040058         324093                  0   
...       ...          ...          ...            ...                ...   
194341   2022        77047        72379           4668                  0   
194342   2022        77065        72185           4880                  0   
194343   2022        77023        71990           5033                  0   
194344   2022        76942        71942           5000                  0   
194345   2023        77249        72170           5079                  0   

        Seasonally Adjusted (Y/N)_new   Area Type _County  Area Type _MSA  \
0                                    0                  0               0   
1                                    1                  0               0   
2                                    0                  1               0   
3                                    1                  1               0   
4                                    0                  0               0   
...                                ...                ...             ...   
194341                               1                  0               1   
194342                               1                  0               1   
194343                               1                  0               1   
194344                               1                  0               1   
194345                               1                  0               1   

        Area Type _Metro Division  Area Type _Metropolitan Area  ...  \
0                               0                             0  ...   
1                               0                             0  ...   
2                               0                             0  ...   
3                               0                             0  ...   
4                               0                             1  ...   
...                           ...                           ...  ...   
194341                          0                             0  ...   
194342                          0                             0  ...   
194343                          0                             0  ...   
194344                          0                             0  ...   
194345                          0                             0  ...   

        Month_December  Month_February  Month_January  Month_July  Month_June  \
0                    0               0              1           0           0   
1                    0               0              1           0           0   
2                    0               0              1           0           0   
3                    0               0              1           0           0   
4                    0               0              1           0           0   
...                ...             ...            ...         ...         ...   
194341               0               0              0           0           0   
194342               0               0              0           0           0   
194343               0               0              0           0           0   
194344               1               0              0           0           0   
194345               0               0              1           0           0   

        Month_March  Month_May  Month_November  Month_October  Month_September  
0                 0          0               0              0                0  
1                 0          0               0              0                0  
2                 0          0               0              0                

In [54]:
X = EncodedData.drop('Unemployment Rate',axis=1) #Splitting the independent variables from the target variable
y = EncodedData['Unemployment Rate']  #Target variable
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 14)  #Splitting the 30% of data to test set and the rest 70% to train
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((136042, 1472), (58304, 1472), (136042,), (58304,))

## Scaling the train data set and using the fitted scaler to transform the training and test data.

In [55]:
scaler = MinMaxScaler()
#Fit = scaler.fit(X_train) 

X_train = scaler.fit_transform(X_train)
X_test = scaler.fit_transform(X_test) 

## Checking for the imbalance in the target value

In [56]:
y_train.value_counts()

0    135768
1       274
Name: Unemployment Rate, dtype: int64

As per the above result we can see that in the total of 194346 almost there are only 274 1s. I will proceed by undersampling the dataset.

In [58]:
Undersample = RandomUnderSampler(sampling_strategy ='majority')
X_train,y_train = Undersample.fit_resample(X_train, y_train)

## Saving data for further running the models

In [60]:
X_train = pd.DataFrame(X_train)
X_train

0             1             2         3     4     5     6     7     \
0    0.744681  3.847962e-03  3.644807e-03  0.002232   0.0   0.0   0.0   0.0   
1    0.787234  3.561792e-03  3.411029e-03  0.001865   0.0   0.0   0.0   0.0   
2    0.851064  5.876702e-04  5.791311e-04  0.000233   0.0   0.0   0.0   0.0   
3    0.787234  1.630146e-02  1.498834e-02  0.012291   0.0   0.0   0.0   0.0   
4    0.851064  2.871919e-03  2.715009e-03  0.001699   0.0   0.0   0.0   0.0   
..        ...           ...           ...       ...   ...   ...   ...   ...   
543  0.957447  2.044070e-06  1.062626e-06  0.000007   0.0   0.0   0.0   0.0   
544  0.723404  1.533053e-06  5.313129e-07  0.000003   0.0   0.0   0.0   0.0   
545  0.936170  5.621193e-06  2.125252e-06  0.000023   0.0   0.0   0.0   0.0   
546  0.936170  6.132210e-05  3.187878e-05  0.000200   0.0   0.0   0.0   0.0   
547  0.893617  5.110175e-07  0.000000e+00  0.000003   0.0   0.0   0.0   0.0   

     8     9     ...  1462  1463  1464  1465  1466  1467  1468  1469  1470  \
0     0.0   0.0  ...   0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0   
1     0.0   0.0  ...   0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0   
2     0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
3     1.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0   
4     0.0   0.0  ...   0.0   0.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   
..    ...   ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
543   0.0   0.0  ...   0.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
544   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
545   0.0   0.0  ...   0.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
546   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0   
547   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   

     1471  
0     0.0  
1     0.0  
2     0.0  
3     0.0  
4     0.0  
..    ...  
543   0.0  
544   0.0  
545   0.0  
546   0.0  
547   1.0  

[548 rows x 1472 columns]

In [61]:
X_train.to_csv('C:/Users/Srinidhi/Documents/USF/Data_Science_Programming/week7_assignment/X_train.csv', index=False)

In [62]:
y_train = pd.DataFrame(y_train)
y_train.to_csv('C:/Users/Srinidhi/Documents/USF/Data_Science_Programming/week7_assignment/y_train.csv', index=False)

In [63]:
X_test = pd.DataFrame(X_test)
X_test.to_csv('C:/Users/Srinidhi/Documents/USF/Data_Science_Programming/week7_assignment/X_test.csv', index=False)

In [64]:
y_test = pd.DataFrame(y_test)
y_test.to_csv('C:/Users/Srinidhi/Documents/USF/Data_Science_Programming/week7_assignment/y_test.csv', index=False)